source/agora/common/TransactionData.d

```diff
@@ -0,0 +1,284 @@
+
+    Contains a storage class consisting of a maximum size fixed byte array.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module agora.common.TransactionData;
+
+import agora.common.Hash;
+import agora.common.Serializer;
+
+static import std.ascii;
+import std.conv: parse;
+import std.algorithm.iteration : each, map;
+import std.range;
+import std.utf;
+
+
+public struct TransactionData
+{
+    invariant()
+    {
+        assert(length <= MAX_SIZE);
+    }
+
+    /// The maximum size
+    public static immutable MAX_SIZE = 1024;
+
+    /// The byte array of transaction data
+    public ubyte[MAX_SIZE] data;
+
+    /// The number of bytes of transaction data
+    public size_t length;
+
+    /// If the input data is larger than MAX_SIZE, this value is true, if not false.
+    public bool is_oversize;
+
+    /***************************************************************************
+
+        Create a TransactionData from binary data
+
+        Params:
+            bin  = Binary data to store in this `TransactionData`
+
+    ***************************************************************************/
+
+    public this (scope const ubyte[] bin) pure nothrow @nogc @safe
+    {
+        if (bin.length > MAX_SIZE)
+            this.is_oversize = true;
+        else
+            this.is_oversize = false;
+
+        size_t offset;
+        foreach (elem; bin)
+        {
+            this.data[offset++] = elem;
+            if (offset == MAX_SIZE)
+                break;
+        }
+        this.length = offset;
+    }
+
+    /***************************************************************************
+
+        Create a TransactionData from an hexadecimal string representation
+
+        Params:
+            hexstr = String representation of the binary data in base 16.
+                     The hexadecimal prefix (0x) is optional.
+                     Can be upper or lower case.
+
+    ***************************************************************************/
+
+    public this (scope const(char)[] hexstr) pure nothrow @nogc @safe
+    {
+        if (hexstr[0] == '0' && (hexstr[1] == 'x' || hexstr[1] == 'X'))
+            hexstr = hexstr[2 .. $];
+
+        if (hexstr.length / 2 > MAX_SIZE)
+            this.is_oversize = true;
+        else
+            this.is_oversize = false;
+
+        auto range = hexstr.byChar.map!(std.ascii.toLower!(char)).map!(parseHex);
+        size_t offset;
+        foreach (chunk; range.chunks(2))
+        {
+            (() @trusted {
+                this.data[offset++] = cast(ubyte)((chunk[0] << 4) + chunk[1]);
+            })();
+            if (offset == MAX_SIZE)
+                break;
+        }
+        this.length = offset;
+    }
```

Comment:
Why do we need an explicit constructor for string => this struct ?

Reply:
This is to process hexadecimal strings when JSON serialized and deserialized.

Reply:
But you don't have to do any of this. Just add a `ubyte[] payload` to the `Transaction`. Vibe.d already does all the serialization/deserialization.